### PR TITLE
[SPARK-44254][SQL] Move QueryExecutionErrors that used by DataType to sql/api as DataTypeErrors

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Moving some QueryExecutionErrors that are used by data types to `sql/api` and name those as DataType erros so that DataType can use those if DataType only stay in `sql/api` module.

### Why are the changes needed?

Towards a simpler DataType interface.

### Does this PR introduce _any_ user-facing change?

No
 
### How was this patch tested?

Existing test

### CODE REVIEW

**Comment by amaliujia** [comment]
@cloud-fan @hvanhovell

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 1]
```diff
@@ -24,6 +24,7 @@ import scala.util.Try
 import org.apache.spark.annotation.Unstable
 import org.apache.spark.sql.catalyst.trees.SQLQueryContext
 import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.errors.SqlApiQueryExecutionErrors
```
Comment (panbingkun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:27]: Do other similar items need to be moved? eg:
QueryExecutionErrors.unscaledValueTooLargeForPrecisionError(this, precision, scale)
It seems that it is only called in `DataType`.

File: sql/api/pom.xml [hunk 2]
```diff
@@ -40,6 +40,11 @@
             <artifactId>spark-common-utils_${scala.binary.version}</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-unsafe_${scala.binary.version}</artifactId>
```
Comment (MaxGekk) [inline sql/api/pom.xml:45]: Which one requires this?

File: sql/api/src/main/scala/org/apache/spark/sql/errors/SqlApiQueryExecutionErrors.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+import org.apache.spark.{SparkArithmeticException, SparkException, SparkRuntimeException, SparkUnsupportedOperationException}
+import org.apache.spark.unsafe.types.UTF8String
+
+/**
+ * Object for grouping error messages from (most) exceptions thrown during query execution.
+ * This does not include exceptions thrown during the eager execution of commands, which are
+ * grouped into [[QueryCompilationErrors]].
+ */
+private[sql] object SqlApiQueryExecutionErrors {
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/errors/SqlApiQueryExecutionErrors.scala:27]: `DataTypeErrors`?

File: sql/api/pom.xml (hunk 2 above)
Comment (amaliujia) [reply sql/api/pom.xml:45]: `def outOfDecimalTypeRangeError(str: UTF8String): SparkArithmeticException`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 1 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:27]: For those we haven't moved, they are using `SQLConf` inside and we are still discussing how to better deal with`SQLConf` now.

File: sql/api/src/main/scala/org/apache/spark/sql/errors/SqlApiQueryExecutionErrors.scala (hunk 3 above)
Comment (amaliujia) [reply sql/api/src/main/scala/org/apache/spark/sql/errors/SqlApiQueryExecutionErrors.scala]: done

File: sql/api/pom.xml (hunk 2 above)
Comment (amaliujia) [reply sql/api/pom.xml:45]: `UTF8String` stays into `spark-unsafe`.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@amaliujia Could you adjust PR title and description according to your recent commit, please.

**Comment by amaliujia** [comment]
@MaxGekk tried to update title and description.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @amaliujia and @cloud-fan @panbingkun for review.

**Pull request closed without merging**